Readme.md
[![Build Status](https://img.shields.io/circleci/project/segmentio/nightmare/master.svg)](https://circleci.com/gh/segmentio/nightmare)
[![Join the chat at https://gitter.im/rosshinkley/nightmare](https://badges.gitter.im/rosshinkley/nightmare.svg)](https://gitter.im/rosshinkley/nightmare?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

# Nightmare

Nightmare is a high-level browser automation library from [Segment](https://segment.com).

The goal is to expose a few simple methods that mimic user actions (like `goto`, `type` and `click`), with an API that feels synchronous for each block of scripting, rather than deeply nested callbacks. It was originally designed for automating tasks across sites that don't have APIs, but is most often used for UI testing and crawling.

Under the covers it uses [Electron](http://electron.atom.io/), which is similar to [PhantomJS](http://phantomjs.org/) but roughly [twice as fast](https://github.com/segmentio/nightmare/issues/484#issuecomment-184519591) and more modern. 

**⚠️ Security Warning:** We've implemented [many](https://github.com/segmentio/nightmare/issues/1388) of the security recommendations [outlined by Electron](https://github.com/electron/electron/blob/master/docs/tutorial/security.md) to try and keep you safe, but undiscovered vulnerabilities may exist in Electron that could allow a malicious website to execute code on your computer. Avoid visiting untrusted websites.

**🛠 Migrating to 3.x:** You'll want to check out [this issue](https://github.com/segmentio/nightmare/issues/1396) before upgrading. We've worked hard to make improvements to nightmare while limiting the breaking changes and there's a good chance you won't need to do anything.

[Niffy](https://github.com/segmentio/niffy) is a perceptual diffing tool built on Nightmare. It helps you detect UI changes and bugs across releases of your web app.

[Daydream](https://github.com/segmentio/daydream) is a complementary chrome extension built by [@stevenmiller888](https://github.com/stevenmiller888) that generates Nightmare scripts for you while you browse.

Many thanks to [@matthewmueller](https://github.com/matthewmueller) and [@rosshinkley](https://github.com/rosshinkley) for their help on Nightmare.

* [Examples](#examples)
  * [UI Testing Quick Start](https://segment.com/blog/ui-testing-with-nightmare/)
  * [Perceptual Diffing with Niffy & Nightmare](https://segment.com/blog/perceptual-diffing-with-niffy/)
* [API](#api)
  * [Set up an instance](#nightmareoptions)
  * [Interact with the page](#interact-with-the-page)
  * [Extract from the page](#extract-from-the-page)
  * [Cookies](#cookies)
  * [Proxies](#proxies)
  * [Promises](#promises)
  * [Extending Nightmare](#extending-nightmare)
* [Usage](#usage)
* [Debugging](#debugging)
* [Additional Resources](#additional-resources)

## Examples

Let's search on DuckDuckGo:

```js
const Nightmare = require('nightmare')
const nightmare = Nightmare({ show: true })

nightmare
  .goto('https://duckduckgo.com')
  .type('#search_form_input_homepage', 'github nightmare')
  .click('#search_button_homepage')
  .wait('#r1-0 a.result__a')
  .evaluate(() => document.querySelector('#r1-0 a.result__a').href)
  .end()
  .then(console.log)
  .catch(error => {
    console.error('Search failed:', error)
  })
```

You can run this with:

```shell
npm install --save nightmare
node example.js
```

Or, let's run some mocha tests:

```js
const Nightmare = require('nightmare')
const chai = require('chai')
const expect = chai.expect

describe('test duckduckgo search results', () => {
  it('should find the nightmare github link first', function(done) {
    this.timeout('10s')

    const nightmare = Nightmare()
    nightmare
      .goto('https://duckduckgo.com')
      .type('#search_form_input_homepage', 'github nightmare')
      .click('#search_button_homepage')
      .wait('#links .result__a')
      .evaluate(() => document.querySelector('#links .result__a').href)
      .end()
      .then(link => {
        expect(link).to.equal('https://github.com/segmentio/nightmare')
        done()
      })
  })
})
```

You can see examples of every function [in the tests here](https://github.com/segmentio/nightmare/blob/master/test/index.js).

To get started with UI Testing, check out this [quick start guide](https://segment.com/blog/ui-testing-with-nightmare).

### To install dependencies

```
npm install
```

### To run the mocha tests

```
npm test
```

### Node versions

Nightmare is intended to be run on NodeJS 4.x or higher.

## API

#### Nightmare(options)

Creates a new instance that can navigate around the web. The available options are [documented here](https://github.com/atom/electron/blob/master/docs/api/browser-window.md#new-browserwindowoptions), along with the following nightmare-specific options.

##### waitTimeout (default: 30s)

Throws an exception if the `.wait()` didn't return `true` within the set timeframe.

```js
const nightmare = Nightmare({
  waitTimeout: 1000 // in ms
})
```

##### gotoTimeout (default: 30s)

Throws an exception if the `.goto()` didn't finish loading within the set timeframe. Note that, even though `goto` normally waits for all the resources on a page to load, a timeout exception is only raised if the DOM itself has not yet loaded.

```js
const nightmare = Nightmare({
  gotoTimeout: 1000 // in ms
})
```

##### loadTimeout (default: infinite)

Forces Nightmare to move on if a page transition caused by an action (eg, `.click()`) didn't finish within the set timeframe. If `loadTimeout` is shorter than `gotoTimeout`, the exceptions thrown by `gotoTimeout` will be suppressed.

```js
const nightmare = Nightmare({
  loadTimeout: 1000 // in ms
})
```

##### executionTimeout (default: 30s)

The maximum amount of time to wait for an `.evaluate()` statement to complete.

```js
const nightmare = Nightmare({
  executionTimeout: 1000 // in ms
})
```

##### paths

The default system paths that Electron knows about. Here's a list of available paths: https://github.com/atom/electron/blob/master/docs/api/app.md#appgetpathname

You can overwrite them in Nightmare by doing the following:

```js
const nightmare = Nightmare({
  paths: {
    userData: '/user/data'
  }
})
```

##### switches

The command line switches used by the Chrome browser that are also supported by Electron. Here's a list of supported Chrome command line switches:
https://github.com/atom/electron/blob/master/docs/api/chrome-command-line-switches.md

```js
const nightmare = Nightmare({
  switches: {
    'proxy-server': '1.2.3.4:5678',
    'ignore-certificate-errors': true
  }
})
```

##### electronPath

The path to the prebuilt Electron binary. This is useful for testing on different versions of Electron. Note that Nightmare only supports the version on which this package depends. Use this option at your own risk.

```js
const nightmare = Nightmare({
  electronPath: require('electron')
})
```

##### dock (OS X)

A boolean to optionally show the Electron icon in the dock (defaults to `false`). This is useful for testing purposes.

```js
const nightmare = Nightmare({
  dock: true
})
```

##### openDevTools

Optionally shows the DevTools in the Electron window using `true`, or use an object hash containing `mode: 'detach'` to show in a separate window. The hash gets passed to [`contents.openDevTools()`](https://github.com/electron/electron/blob/master/docs/api/web-contents.md#contentsopendevtoolsoptions) to be handled. This is also useful for testing purposes. Note that this option is honored only if `show` is set to `true`.

```js
const nightmare = Nightmare({
  openDevTools: {
    mode: 'detach'
  },
  show: true
})
```

##### typeInterval (default: 100ms)

How long to wait between keystrokes when using `.type()`.

```js
const nightmare = Nightmare({
  typeInterval: 20
})
```

##### pollInterval (default: 250ms)

How long to wait between checks for the `.wait()` condition to be successful.

```js
const nightmare = Nightmare({
  pollInterval: 50 //in ms
})
```

##### maxAuthRetries (default: 3)

Defines the number of times to retry an authentication when set up with `.authenticate()`.

```js
const nightmare = Nightmare({
  maxAuthRetries: 3
})
```

#### certificateSubjectName

A string to determine the client certificate selected by electron. If this options is set, the [`select-client-certificate`](https://github.com/electron/electron/blob/master/docs/api/app.md#event-select-client-certificate) event will be set to loop through the certificateList and find the first certificate that matches `subjectName` on the electron [`Certificate Object`](https://electronjs.org/docs/api/structures/certificate).

```js
const nightmare = Nightmare({
  certificateSubjectName: 'tester'
})
```

#### .engineVersions()

Gets the versions for Electron and Chromium.

#### .useragent(useragent)

Sets the `useragent` used by electron.

#### .authentication(user, password)

Sets the `user` and `password` for accessing a web page using basic authentication. Be sure to set it before calling `.goto(url)`.

#### .end()

Completes any queue operations, disconnect and close the electron process. Note that if you're using promises, `.then()` must be called after `.end()` to run the `.end()` task. Also note that if using an `.end()` callback, the `.end()` call is equivalent to calling `.end()` followed by `.then(fn)`. Consider:

```js
nightmare
  .goto(someUrl)
  .end(() => 'some value')
  //prints "some value"
  .then(console.log)
```

#### .halt(error, done)

Clears all queued operations, kills the electron process, and passes error message or 'Nightmare Halted' to an unresolved promise. Done will be called after the process has exited.

### Interact with the Page

#### .goto(url[, headers])

Loads the page at `url`. Optionally, a `headers` hash can be supplied to set headers on the `goto` request.

When a page load is successful, `goto` returns an object with metadata about the page load, including:

* `url`: The URL that was loaded
* `code`: The HTTP status code (e.g. 200, 404, 500)
* `method`: The HTTP method used (e.g. "GET", "POST")
* `referrer`: The page that the window was displaying prior to this load or an empty string if this is the first page load.
* `headers`: An object representing the response headers for the request as in `{header1-name: header1-value, header2-name: header2-value}`

If the page load fails, the error will be an object with the following properties:

* `message`: A string describing the type of error
* `code`: The underlying error code describing what went wrong. Note this is NOT the HTTP status code. For possible values, see https://code.google.com/p/chromium/codesearch#chromium/src/net/base/net_error_list.h
* `details`: A string with additional details about the error. This may be null or an empty string.
* `url`: The URL that failed to load

Note that any valid response from a server is considered “successful.” That means things like 404 “not found” errors are successful results for `goto`. Only things that would cause no page to appear in the browser window, such as no server responding at the given address, the server hanging up in the middle of a response, or invalid URLs, are errors.

You can also adjust how long `goto` will wait before timing out by setting the [`gotoTimeout` option](#gototimeout-default-30s) on the Nightmare constructor.

#### .back()

Goes back to the previous page.

#### .forward()

Goes forward to the next page.

#### .refresh()

Refreshes the current page.

#### .click(selector)

Clicks the `selector` element once.

#### .mousedown(selector)

Mousedowns the `selector` element once.

#### .mouseup(selector)

Mouseups the `selector` element once.

#### .mouseover(selector)

Mouseovers the `selector` element once.

#### .mouseout(selector)

Mouseout the `selector` element once.

#### .type(selector[, text])


History.md
# 3.0.1 / 2018-03-29

* bump electron to fix #1424

# 3.0.0 / 2018-03-02

* BREAKING: remove window.\_\_nightmare.ipc to resolve the major security issues (#1390)
* BREAKING: properly serialize error values (#1391)
* added linting, formatting and a git pre-commit hook (#1386)
* Added ability to specify the client certificate selected by electron (#1339)
* adding selector to the timeout error in .wait() (#1381)
* Pin Electron version (#1270)
* Fix error on preload (#1247)
* Add mouseout action to complement mouseover (#1238)
* fix problems that are rejected when adding child actions with objects. (#1093)
* Set mouse position for mouse events (#1077)
* Repaired support for multiple timeouts in FrameManager (#945)

# 2.10.0 / 2017-02-23

* Remove redundant docs for 'log' event from README
* changed some `var` declarations to `const`
* replace the 404 link with valid link
* added Promise override tests
* added docs for new Promise override features
* added ability to override internal Promise library

# 2.9.1 / 2017-01-02

* Minor touchups to key press documentation
* Link to Electron documentation updated
* Updates speed information on the readme
* Swaps Yahoo example out for a faster DuckDuckGo example
* Fixes an issue where `nightmare` may be undefined in the browser at injection time
* Changes screenshot rendering to use debugger API instead of forcing a DOM change

# 2.9.0 / 2016-12-17

* Prevents unload dialogs, allowing Nightmare to end gracefully
* `.end(fn)` now uses `.then()` under the covers
* **Possibly breaking change:** Nightmare will now default to using a non-persistent partition. Data between executions of Nightmare will no longer be saved.
* Adds `.mouseup()` action
* Fixes several typos/copy-paste errors in the readme, as well as clarifying error-first callbacks
* Adds `.path()` to get the URL's route (as opposed to the fully-qualified URL)

# 2.8.1 / 2016-10-20

* Fixes parsing issues with arguments to `evaluate_now`
* Upgrades to Electron 1.4.4

# 2.8.0 / 2016-10-20

* Fixes a missing semicolon in the first readme example
* Fixes a reference error inside `.wait()` when using `node --use_strict`
* Adds missing documentation for `.mouseover()`
* Corrects a typo in the readme
* Removes dependency on `object-assign`
* Adds `.halt()` API to stop Nightmare execution immediately
* Fixes `blur` exception when elements are removed by keyboard events
* **Possibly breaking change:** Changes `.evaluate()` to allow for asynchronous execution. If the `.evaluate()`d function's arity is one less than the passed in parameters, it'll assume the last argument to the function is the callback. If the return value is a thenable, it'll call `then()` to wait for promise fulfillment. Otherwise, the call will behave synchronously as it does now.

# 2.7.0 / 2016-09-05

* Adds `.wait(element, timeout)` to wait for whichever comes first
* `.end()` will now end Electron gracefully instead of issuing a `SIGKILL`
* Touches up readme for `.end()`

# 2.6.1 / 2016-08-08

* Fixes treating provisional load failures as real load failures

# 2.6.0 / 2016-08-02

* Makes the CircleCI badge an SVG
* Adds an option for `.type()` to control time elapsed between keystrokes with `typeInterval`
* Adds `.cookies.clearAll()` to clear all cookies
* Fixes crashing if the Electron process is closed first
* Adds `pollInterval` as an option to control the tick time for `.wait()`
* Forces Nightmare to error on bad HTTP authentication
* Fixes a crash by omitting event data due to circular references
* Adds environment variable forwarding to the Electron process
* Fixes `openDevTools` docs to be more explicit about detaching the devtools tray
* Fixes the link to the preload script

# 2.5.3 / 2016-07-08

* Adds better proxy information to the readme
* Fixes a readme typo
* Updates `ipcRenderer` usage for preload scripts in readme
* Bumps Electron to version 1.2.5

# 2.5.2 / 2016-06-20

* Fixes `Referer` header support
* Removes timeout between keystrokes when using `.type()`
* Checks instance existence when calling `.end()`
* Adds a link to `nightmare-examples`
* Changes `yield` to `.then()` in readme
* Swaps `did-finish-loading` for `did-stop-loading` when waiting for page transitions
* Adds optional `loadTimeout` for server responses that do not end

# 2.5.1 / 2016-06-07

* Bumps Electron dependency to 1.2.1.
* Removes a `sender` workaround
* Moves the start of Electron from the constructor into the queue

# 2.5.0 / 2016-05-27

* adds a timeout to `.goto()` such that pages that load the DOM but never finish are considered successful, otherwise failing, preventing a hang.
* updates the example script and readme file for consistency.
* reports with more helpful messages when the element does not exist when running `.click()`, `.mousedown()` and `.mouseover()`.
* `.coookies.clear()` with no arguments will clear all cookies for the current domain.
* adds Node engine information to package and ensures CircleCI builds and tests against 4.x, 5.x and 6.x.
* removes extranneous `javascript` event listeners upon execution completion or error.
* adds `.once()` and `.removeListener()` for more complete Electron process event handling.

# 2.4.1 / 2016-05-19

* Points invalid test URLs to the `.tld` domain
* Switches javascript templates over to using template strings.
* Adds better switch tests
* Javascript `goto`s now only wait if the main frame is loading
* Allows a Nightmare instance to use `.catch()` without a `.then()`
* Fixes a deprecated IPC inclusion in tests
* `.goto()` rejects with a helpful message when `url` is not provided

# 2.4.0 / 2016-05-05

* adds call safety with IPC callbacks
* adds `.engineVersions()` to get Electron and Chrome versions, as well as Nightmare.version
* changes Yahoo example to use more robust selectors, adds `.catch()`
* adds a check for `runner` arguments

# 2.3.4 / 2016-04-23

* blurs text inputs when finished with `.type()` or `.input()`, including clearing selectors

lib/runner.js
/**
 * Module Dependencies
 */

var parent = require('./ipc')(process)
var electron = require('electron')
var BrowserWindow = electron.BrowserWindow
var defaults = require('deep-defaults')
var join = require('path').join
var sliced = require('sliced')
var renderer = require('electron').ipcMain
var app = require('electron').app
var urlFormat = require('url')
var FrameManager = require('./frame-manager')

// URL protocols that don't need to be checked for validity
const KNOWN_PROTOCOLS = ['http', 'https', 'file', 'about', 'javascript']
// Property for tracking whether a window is ready for interaction
const IS_READY = Symbol('isReady')

/**
 * Handle uncaught exceptions in the main electron process
 */

process.on('uncaughtException', function(e) {
  parent.emit('uncaughtException', e.stack)
})

/**
 * Update the app paths
 */

if (process.argv.length < 3) {
  throw new Error(`Too few runner arguments: ${JSON.stringify(process.argv)}`)
}

var processArgs = JSON.parse(process.argv[2])
var paths = processArgs.paths
if (paths) {
  for (let i in paths) {
    app.setPath(i, paths[i])
  }
}
var switches = processArgs.switches
if (switches) {
  for (let i in switches) {
    app.commandLine.appendSwitch(i, switches[i])
  }
  var chromiumArguments = processArgs.chromiumArguments;
  if (chromiumArguments) {
    for (var i in chromiumArguments) {
      app.commandLine.appendArgument(chromiumArguments[i]);
    }
  }
}

/**
 * Hide the dock
 */

// app.dock is not defined when running
// electron in a platform other than OS X
if (!processArgs.dock && app.dock) {
  app.dock.hide()
}

/**
 * Set the client certificate by subjectName if processArgs.certificateSubjectName is defined
 */

if (processArgs.certificateSubjectName) {
  app.on(
    'select-client-certificate',
    (event, webContents, url, list, callback) => {
      for (var i = 0; i < list.length; i++) {
        if (list[i].subjectName === processArgs.certificateSubjectName) {
          callback(list[i])
          return
        }
      }

      // defaults to first if the subject name is not available
      callback(list[0])
    }
  )
}

/**
 * Listen for the app being "ready"
 */

app.on('ready', function() {
  var win, frameManager, options, closed

  /**
   * create a browser window
   */

  parent.respondTo('browser-initialize', function(opts, done) {
    options = defaults(opts || {}, {
      show: false,
      alwaysOnTop: true,
      webPreferences: {
        preload: join(__dirname, 'preload.js'),
        nodeIntegration: false
      }
    })

    /**
     * Create a new Browser Window
     */

    win = new BrowserWindow(options)
    if (options.show && options.openDevTools) {
      if (typeof options.openDevTools === 'object') {
        win.openDevTools(options.openDevTools)
      } else {
        win.openDevTools()
      }
    }

    /**
     * Window Docs:
     * https://github.com/atom/electron/blob/master/docs/api/browser-window.md
     */

    frameManager = FrameManager(win)

    /**
     * Window options
     */

    win.webContents.setAudioMuted(true)

    /**
     * Pass along web content events
     */

    renderer.on('page', function(_sender /*, arguments, ... */) {
      parent.emit.apply(parent, ['page'].concat(sliced(arguments, 1)))
    })

    renderer.on('console', function(sender, type, args) {
      parent.emit.apply(parent, ['console', type].concat(args))
    })

    win.webContents.on('did-finish-load', forward('did-finish-load'))
    win.webContents.on('did-fail-load', forward('did-fail-load'))
    win.webContents.on(
      'did-fail-provisional-load',
      forward('did-fail-provisional-load')
    )
    win.webContents.on(
      'did-frame-finish-load',
      forward('did-frame-finish-load')
    )
    win.webContents.on('did-start-loading', forward('did-start-loading'))
    win.webContents.on('did-stop-loading', forward('did-stop-loading'))
    win.webContents.on(
      'did-get-response-details',
      forward('did-get-response-details')
    )
    win.webContents.on(
      'did-get-redirect-request',
      forward('did-get-redirect-request')
    )
    win.webContents.on('dom-ready', forward('dom-ready'))
    win.webContents.on('page-favicon-updated', forward('page-favicon-updated'))
    win.webContents.on('new-window', forward('new-window'))
    win.webContents.on('will-navigate', forward('will-navigate'))
    win.webContents.on('crashed', forward('crashed'))
    win.webContents.on('plugin-crashed', forward('plugin-crashed'))
    win.webContents.on('destroyed', forward('destroyed'))
    win.webContents.on(
      'media-started-playing',
      forward('media-started-playing')
    )
    win.webContents.on('media-paused', forward('media-paused'))
    win.webContents.on('close', _e => {
      closed = true
    })

    var loadwatch
    win.webContents.on('did-start-loading', function() {
      if (win.webContents.isLoadingMainFrame()) {
        if (options.loadTimeout) {
          loadwatch = setTimeout(function() {
            win.webContents.stop()
          }, options.loadTimeout)
        }
        setIsReady(false)
      }
    })

    win.webContents.on('did-stop-loading', function() {
      clearTimeout(loadwatch)
      setIsReady(true)
    })

    setIsReady(true)

    done()
  })

  /**
   * Parent actions
   */

  /**
   * goto
   */

  parent.respondTo('goto', function(url, headers, timeout, done) {
    if (!url || typeof url !== 'string') {
      return done('goto: `url` must be a non-empty string')
    }

    var httpReferrer = ''
    var extraHeaders = ''
    for (var key in headers) {
      if (key.toLowerCase() == 'referer') {
        httpReferrer = headers[key]
        continue
      }

      extraHeaders += key + ': ' + headers[key] + '\n'
    }
    var loadUrlOptions = { extraHeaders: extraHeaders }
    httpReferrer && (loadUrlOptions.httpReferrer = httpReferrer)

    if (win.webContents.getURL() == url) {
      done()
    } else {
      var responseData = {}
      var domLoaded = false

      var timer = setTimeout(function() {
        // If the DOM loaded before timing out, consider the load successful.
        var error = domLoaded
          ? undefined
          : {
              message: 'navigation error',
              code: -7, // chromium's generic networking timeout code
              details: `Navigation timed out after ${timeout} ms`,
              url: url
            }
        // Even if "successful," note that some things didn't finish.
        responseData.details = `Not all resources loaded after ${timeout} ms`
        cleanup(error, responseData)
      }, timeout)

      function handleFailure(event, code, detail, failedUrl, isMainFrame) {
        if (isMainFrame) {
          cleanup({
            message: 'navigation error',
            code: code,
            details: detail,
            url: failedUrl || url
          })
        }
      }

      function handleDetails(
        event,
        status,
        newUrl,
        oldUrl,
        statusCode,
        method,
        referrer,
        headers,
        resourceType
      ) {
        if (resourceType === 'mainFrame') {
          responseData = {
            url: newUrl,
            code: statusCode,
            method: method,
            referrer: referrer,
            headers: headers
          }
        }
      }

      function handleDomReady() {
        domLoaded = true
      }

      // We will have already unsubscribed if load failed, so assume success.
      function handleFinish(_event) {
        cleanup(null, responseData)
      }

      function cleanup(error, data) {
        clearTimeout(timer)
        win.webContents.removeListener('did-fail-load', handleFailure)
        win.webContents.removeListener(
          'did-fail-provisional-load',
          handleFailure
        )
        win.webContents.removeListener(
          'did-get-response-details',
          handleDetails
        )
        win.webContents.removeListener('dom-ready', handleDomReady)
        win.webContents.removeListener('did-finish-load', handleFinish)
        setIsReady(true)
        // wait a tick before notifying to resolve race conditions for events
        setImmediate(() => done(error, data))
      }

      // In most environments, loadURL handles this logic for us, but in some
      // it just hangs for unhandled protocols. Mitigate by checking ourselves.
      function canLoadProtocol(protocol, callback) {
        protocol = (protocol || '').replace(/:$/, '')
        if (!protocol || KNOWN_PROTOCOLS.includes(protocol)) {
          return callback(true)
        }
        electron.protocol.isProtocolHandled(protocol, callback)
      }

      function startLoading() {
        // abort any pending loads first
        if (win.webContents.isLoading()) {
          parent.emit('log', 'aborting pending page load')
          win.webContents.once('did-stop-loading', function() {
            startLoading(true)
          })
          return win.webContents.stop()
        }

        win.webContents.on('did-fail-load', handleFailure)
        win.webContents.on('did-fail-provisional-load', handleFailure)
        win.webContents.on('did-get-response-details', handleDetails)
        win.webContents.on('dom-ready', handleDomReady)
        win.webContents.on('did-finish-load', handleFinish)
        win.webContents.loadURL(url, loadUrlOptions)

        // javascript: URLs *may* trigger page loads; wait a bit to see
        if (protocol === 'javascript:') {
          setTimeout(function() {
            if (!win.webContents.isLoadingMainFrame()) {
              done(null, {
                url: url,
                code: 200,
                method: 'GET',
                referrer: win.webContents.getURL(),
                headers: {}
              })
            }
          }, 10)
        }
      }

      var protocol = urlFormat.parse(url).protocol
      canLoadProtocol(protocol, function startLoad(canLoad) {
        if (canLoad) {
          parent.emit(
            'log',
            `Navigating: "${url}",
            headers: ${extraHeaders || '[none]'},
            timeout: ${timeout}`
          )
          return startLoading()
        }

        cleanup({
          message: 'navigation error',
          code: -1000,
          details: 'unhandled protocol',
          url: url
        })
      })
    }
  })

  /**
   * javascript
   */

  parent.respondTo('javascript', function(src, done) {
    var response = (event, response) => {
      renderer.removeListener('error', error)
      renderer.removeListener('log', log)
      done(null, response)
    }

    var error = (event, error) => {
      renderer.removeListener('log', log)
      renderer.removeListener('response', response)
      done(error)
    }

    var log = (event, args) => parent.emit.apply(parent, ['log'].concat(args))

    renderer.once('response', response)
    renderer.once('error', error)
    renderer.on('log', log)

    //parent.emit('log', 'about to execute javascript: ' + src);
    win.webContents.executeJavaScript(src)
  })

  /**
   * css
   */

  parent.respondTo('css', function(css, done) {
    win.webContents.insertCSS(css)
    done()
  })

  /**
   * size
   */

  parent.respondTo('size', function(width, height, done) {
    win.setSize(width, height)
    done()
  })

  parent.respondTo('useragent', function(useragent, done) {
    win.webContents.setUserAgent(useragent)
    done()
  })

  /**
   * type
   */

  parent.respondTo('type', function(value, done) {
    var chars = String(value).split('')

    function type() {
      var ch = chars.shift()
      if (ch === undefined) {
        return done()
      }

      // keydown
      win.webContents.sendInputEvent({
        type: 'keyDown',
        keyCode: ch
      })

      // keypress
      win.webContents.sendInputEvent({
        type: 'char',
        keyCode: ch
      })

      // keyup
      win.webContents.sendInputEvent({
        type: 'keyUp',
        keyCode: ch
      })

      // defer function into next event loop
      setTimeout(type, options.typeInterval)
    }

    // start
    type()
  })

  /**
   * Insert
   */

  parent.respondTo('insert', function(value, done) {
    win.webContents.insertText(String(value))
    done()
  })

  /**
   * screenshot
   */

  parent.respondTo('screenshot', function(path, clip, done) {
    // https://gist.github.com/twolfson/0d374d9d7f26eefe7d38
    var args = [
      function handleCapture(img) {
        done(null, img.toPng())
      }
    ]
    if (clip) args.unshift(clip)
    frameManager.requestFrame(function() {
      win.capturePage.apply(win, args)
    })
  })

  /**
   * html
   */

  parent.respondTo('html', function(path, saveType, done) {
    // https://github.com/atom/electron/blob/master/docs/api/web-contents.md#webcontentssavepagefullpath-savetype-callback
    saveType = saveType || 'HTMLComplete'
    win.webContents.savePage(path, saveType, function(error) {
      done(error)
    })
  })

  /**
   * pdf
   */

  parent.respondTo('pdf', function(path, options, done) {
    // https://github.com/fraserxu/electron-pdf/blob/master/index.js#L98
    options = defaults(options || {}, {
      marginType: 0,
      printBackground: true,
      printSelectionOnly: false,
      landscape: false
    })

    win.webContents.printToPDF(options, function(error, data) {
      if (error) return done(arguments)
      done(null, data)
    })
  })

  /**
   * Get cookies
   */

  parent.respondTo('cookie.get', function(query, done) {
    var details = Object.assign(
      {
        url: win.webContents.getURL()
      },
      query
    )

    parent.emit('log', 'getting cookie: ' + JSON.stringify(details))
    win.webContents.session.cookies.get(details, function(error, cookies) {
      if (error) return done(error)
      done(null, details.name ? cookies[0] : cookies)
    })
  })

  /**
   * Set cookies
   */

  parent.respondTo('cookie.set', function(cookies, done) {
    var pending = cookies.length

    for (var i = 0, cookie; (cookie = cookies[i]); i++) {
      var details = Object.assign(
        {
          url: win.webContents.getURL()
        },
        cookie
      )

      parent.emit('log', 'setting cookie: ' + JSON.stringify(details))
      win.webContents.session.cookies.set(details, function(error) {
        if (error) done(error)
        else if (!--pending) done()
      })
    }
  })

  /**
   * Clear cookie
   */

  parent.respondTo('cookie.clear', function(cookies, done) {
    var url = win.webContents.getURL()
    var getCookies = cb => cb(null, cookies)

    if (cookies.length == 0) {
      getCookies = cb =>
        win.webContents.session.cookies.get({ url: url }, (error, cookies) => {
          cb(error, cookies.map(cookie => cookie.name))
        })
    }

    getCookies((error, cookies) => {
      var pending = cookies.length
      //if no cookies, return
      if (pending == 0) {
        return done()
      }
      parent.emit('log', 'listing params', cookies)

      //for each cookie name in cookies,
      for (var i = 0, cookie; (cookie = cookies[i]); i++) {
        //remove the cookie from the url
        win.webContents.session.cookies.remove(url, cookie, function(error) {
          if (error) done(error)
          else if (!--pending) done()
        })
      }
    })
  })

  /**
   * Clear all cookies
   */

  parent.respondTo('cookie.clearAll', function(done) {
    win.webContents.session.clearStorageData(
      {
        storages: ['cookies']
      },
      done
    )
  })

  /**
   * Add custom functionality
   */

  parent.respondTo('action', function(name, fntext, done) {
    var fn = new Function(
      'with(this){ parent.emit("log", "adding action for ' +
        name +
        '"); return ' +
        fntext +
        '}'
    ).call({
      require: require,
      parent: parent
    })
    fn(name, options, parent, win, renderer, function(error) {
      done(error)
    })
  })

  /**
   * Continue
   */

  parent.respondTo('continue', function(done) {
    if (isReady()) {
      done()
    } else {
      parent.emit('log', 'waiting for window to load...')
      win.once('did-change-is-ready', function() {
        parent.emit('log', 'window became ready: ' + win.webContents.getURL())
        done()
      })
    }
  })

  /**
   * Authentication
   */

  var loginListener
  parent.respondTo('authentication', function(login, password, done) {
    var currentUrl
    var tries = 0
    if (loginListener) {
      win.webContents.removeListener('login', loginListener)
    }

    loginListener = function(webContents, request, authInfo, callback) {
      tries++
      parent.emit('log', `authenticating against ${request.url}, try #${tries}`)
      if (currentUrl != request.url) {
        currentUrl = request.url
        tries = 1
      }

      if (tries >= options.maxAuthRetries) {
        parent.emit('die', 'problem authenticating, check your credentials')
      } else {
        callback(login, password)
      }
    }
    win.webContents.on('login', loginListener)

    done()
  })

  /**
   * Kill the electron app
   */

  parent.respondTo('quit', function(done) {
    app.quit()
    done()
  })

  /**
   * Send "ready" event to the parent process
   */

  parent.emit('ready', {
    electron: process.versions['electron'],
    chrome: process.versions['chrome']
  })

  /**
   * Check whether the window is ready for interaction
   */
  function isReady() {
    return win[IS_READY]
  }

  /**
   * Set whether the window is ready for interaction
   */
  function setIsReady(ready) {
    ready = !!ready
    if (ready !== win[IS_READY]) {
      win[IS_READY] = ready
      win.emit('did-change-is-ready', ready)
    }
  }

  /**
   * Forward events
   */

  function forward(name) {
    return function(_event) {
      // NOTE: the raw Electron event used to be forwarded here, but we now send
      // an empty event in its place -- the raw event is not JSON serializable.
      if (!closed) {
        parent.emit.apply(parent, [name, {}].concat(sliced(arguments, 1)))
      }
    }
  }
})
